lib/charms/tls_certificates_interface/v1/tls_certificates.py

```diff
@@ -0,0 +1,594 @@
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Library for the tls-certificates relation.
+
+This library contains the Requires and Provides classes for handling the tls-certificates
+interface.
+
+## Getting Started
+From a charm directory, fetch the library using `charmcraft`:
+
+```shell
+charmcraft fetch-lib charms.tls_certificates_interface.v1.tls_certificates
+```
+
+Add the following library to the charm's `requirements.txt` file:
+- jsonschema
+- cryptography
+
+Add the following section to the charm's `charmcraft.yaml` file:
+```yaml
+parts:
+  charm:
+    build-packages:
+      - libffi-dev
+      - libssl-dev
+      - rustc
+      - cargo
+```
+
+### Provider charm
+Example:
+```python
+
+import logging
+
+from ops.charm import CharmBase, InstallEvent
+from ops.framework import StoredState
+from ops.main import main
+from ops.model import ActiveStatus
+
+from charms.tls_certificates_interface.v1.tls_certificates import (
+    CertificateRequestEvent,
+    TLSCertificatesProvides,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class ExampleProviderCharm(CharmBase):
+
+    _stored = StoredState()
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.certificates = TLSCertificatesProvides(self, "certificates")
+        self._stored.set_default(ca_private_key=b"", ca_certificate=b"")
```

Comment:
This (I assume) implies that a given provider charm only ever provides certs from a single CA? Might be worth clarifying in the docs up top alongside where you explain the role :)

Reply:
I'd maybe also note that charmers should consider whether `StoredState` is the right place to store the private key, but I appreciate this is just an example!

Reply:
Fixed, now using peer relation data